common/clock/ratelimiter.go

```diff
@@ -0,0 +1,418 @@
+		// Wait blocks until one token is available (and consumes it), or it returns
+		// an error and does not consume any tokens if:
+		//   - context is canceled
+		//   - a token will never become available (burst == 0)
+		//   - a token is not expected to become available in time
+		//     (short deadline and/or many pending reservations or waiters)
+		//
+		// If an error is returned, regardless of the reason, you may NOT proceed
+		// with the limited event you were waiting for.
+		Wait(ctx context.Context) error
+	}
+
+	// Reservation is a simplified and test-friendly version of [golang.org/x/time/rate.Reservation]
+	// that only allows checking if it is successful or not, and (possibly) returning unused tokens.
+	//
+	// This Reservation MUST have Allow() checked and marked as Used(true/false)
+	// as soon as possible, or an unused token is unlikely to be returned.
+	//
+	// Recommended use is either by using defer:
+	//
+	// 	func allow() (allowed bool) {
+	//     	r := ratelimiter.Reserve()
+	// 		// defer in a closure to read the most-recent value, not the value when deferred
+	// 		defer func() { r.Used(allowed) }()
+	// 		allowed = r.Allow()
+	// 		if !allowed {
+	//			return false // also marks the reservation as not-used
+	// 		}
+	// 		if somethingElseAllowsIt() {
+	// 			// allow the event, and mark the reservation as used so the token is consumed
+	//			return true
+	// 		}
+	// 		// cancel the reservation and possibly return the token
+	// 		return false
+	// 	}
+	//
+	// Or directly calling Used when appropriate:
+	//
+	// 	func allow() bool {
+	//     	r := ratelimiter.Reserve()
+	// 		allowed = r.Allow()
+	// 		if !allowed {
+	// 			r.Used(false) // mark the reservation as not-used
+	//			return false
+	// 		}
+	// 		if somethingElseAllowsIt() {
+	// 			// allow the event, and mark the reservation as used
+	// 			r.Used(true)
+	//			return true
+	// 		}
+	// 		r.Used(false) // mark the reservation as not-used, and restore the unused token
+	// 		return false
+	// 	}
+	Reservation interface {
+		// Allow returns true if a request should be allowed.
+		//
+		// This may be called at any time and it will return the same value each time,
+		// but it becomes less correct as time passes, so you are expected to call it ~immediately.
+		//
+		// As soon as possible, also call Used(true/false) to consume / return the reserved token.
+		//
+		// This is equivalent to `OK() == true && DelayFrom(reservationTime) == 0`
+		// with [rate.Reservation].
+		Allow() bool
+
+		// Used marks this Reservation as either used or not-used, and it must be called
+		// once on every Reservation:
+		//  - If true, the event is assumed to be allowed, and the Ratelimiter token
+		//    will remain consumed.
+		//  - If false, the Reservation will be rolled back, possibly restoring a Ratelimiter token.
+		//    This is equivalent to calling Cancel() on a [rate.Reservation], but it is more likely
+		//    to return the token.
+		Used(wasUsed bool)
```

Comment:
dumb question: what's the intent one might try to convey by specifying that a reservation was used, even when it wasn't allowed? What does that do to the underlying limiter?

Reply:
In this implementation: nothing, it's always a no-op.

In the global ratelimiter thing that led to all this: so you have an "accepted/rejected" event that can be monitored, for counting usage.

With just "Reserve -> Cancel", you have to assume all reserves are "used" immediately, and subtract 1 if cancel is called.  And then you may have to deal with counts going negative briefly.  
With something called in *all* cases, that's not a concern, just defer the count until it's marked as used-or-not.

I suspect it'll also be useful in tests, as this way you always `EXPECT` a call (but the arg varies) rather than "it might be called or it might not".

Reply:
Ah, you probably meant with the real limiter / a somewhat different question.

We (currently and in this wrapper) consider a reservation "allowed" if it is `OK()` (possible to reserve, literally just "N <= burst") and `Delay() == 0` (the limiter had a full token available *right now*, which has already been claimed).

Delay can be positive, if there wasn't a full token in the bucket.  You can take advantage of this if you wanted to reserve a token and wait for it to become available.  
At that point, reservations are essentially identical to `Wait()` _but_ you can tell ahead of time how long you would have to wait, instead of it being a black box that you can give up on (cancel the context) but not predict.

When you get one of these positive-delay reservations, _a token has still been removed_.  The limiter goes from 0.5 tokens to -0.5 tokens.  Reserve another and you'll be at -1.5 tokens, and it'll be 2.5 ratelimiter-periods before any `Allow()` call can succeed, as it refills the bucket from negative values.  But you can continue `Reserve()`ing (or `Wait()`ing) as much as you like, and push that to-be-allowed-call backlog out arbitrarily far.

Reply:
> *: ... I'm not actually sure this is reasonable, now that I think about it. I suspect it can allow arbitrarily large bursts.

Ah, no, that's handled.  It may take me a while to fully understand the details, but it's tracking "reserves since reserved" and not exceeding that value, so carefully-timed reserve+cancel interactions don't seem to allow scheduling a bunch of reserved events at the same time.